User invites

begedin opened the pull request:
# What's in this PR?

This PR adds API support for `UserInvite` creation and claims.

The Issue this is referring to specifies that we should be adding an `InvitedUser` model, but `UserInvite` seems more consistent in naming, so I opted for that.

## Creation of user  invites

- we have a new controller for this, as well as a new endpoint
- we differentiate between a plain invite and an invite created to invite a user to a project
- when creating a plain invite, the user is authorized to do so if they are the inviter, with no other conditions
- when creating a project invite, the user needs to be an inviter, and additionally, they need to be of a high enough membership in the specified project, in order to invite a user with a specified role
  - pending users and contributors (?) can't do anything
  - admins can invite other pending (not useful) and contributor users
  - owners can do the same as admins and also invite other admin users
- during creation of the invite, we require an email and also validate that a user account with that email doesn't exist already

## Claiming of user invites

This is done through the user controller create action. The user creation was extracted into `Accounts.create/1` and a new, special clause matches against a payload holding an `"invite_id"` key, to differentiate from normal account creation.

When claiming the invite we

 - try to load the invite by specified id - can fail and renders a 404 if it does
 - create the user (this can fail with a validation error if the user has been created regularly since the invite was created)
  - create a project user if a project and role are specified by the invite
  - associate invite with created user through `invitee`

The user controller is where we track any claimed invites.

## Suggested issues to create 

- [ ] Add an extra step to normal account creation, which claims all invites associated with the created user's email
- [ ] Add the same extra step to invite claim account creation, which claims all invites other than the one currently being claimed.
- [ ] Add support for inviting existing users to project

## References

Closes #1350

begedin committed:
Add Accounts.create_invite

- Add UserInvite schema and model
- Add Accounts.UserInvites subcontext
- Add Accounts.create_invite tests

begedin committed:
Add Accounts.claim_invite/1

- Add Accounts.claim_invite/1 tests
- Add validation step to Accounts.create_invite, to ensure invitee is not a member of the project already

begedin committed:
Switch Accounts.claim_invite to accept a params map

joshsmith commented on lib/code_corps/model/user_invite.ex:
```diff
@@ -0,0 +1,17 @@
+
+  schema "user_invites" do
+    field :email, :string, null: false
+    field :role, :string, default: "contributor"
```
This shouldn't have a default because it's possible for an invite to _not_ be for a project.

joshsmith commented on priv/repo/migrations/20171228163712_add_user_invites.exs:
```diff
@@ -0,0 +1,17 @@
+    create table(:user_invites) do
+      add :email, :string, null: false
+      add :name, :string, null: true
+      add :role, :string, null: true, default: "contributor"
```
Same here re: default.

joshsmith commented on priv/repo/structure.sql:
```diff
@@ -2,8 +2,8 @@
--- Dumped from database version 10.1
--- Dumped by pg_dump version 10.1
+-- Dumped from database version 10.0
+-- Dumped by pg_dump version 10.0
```
Can we see if there's a way to avoid the db version numbering while still getting the benefits of the structure.sql?

begedin replied:
I'm not seeing any documented way of doing so with `mix ecto.dump`

joshsmith commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,100 @@
+    %UserInvite{}
+    |> Changeset.cast(params, [:email, :name, :role, :inviter_id, :project_id])
+    |> Changeset.validate_required([:email, :inviter_id])
+    |> Changeset.validate_inclusion(:role, ~w(contributor admin owner))
```
Ideally we would source the role types from the `ProjectUser`.

joshsmith commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,100 @@
+
+        if count > 0 do
+          changeset |> Changeset.add_error(:email, "Already associated with a project member")
+        else
```
I think the requirement wasn't quite clear here: we want to be sure that the email is not associated with a `User`.

begedin replied:
Several questions then

- will a project admin/owner ever be able to invite an existing codecorps user to their project?
  - would that not create a `UserInvite` and create something else? Maybe a `ProjectUser` record with a status of `"invited"`?

- what happens in the following scenario:
  - invite is created for a "mike@mail.com"
  - later, independently, a user registers with that email
  - even later, that same user tries to claim the invite

joshsmith commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  @spec claim_invite(map) :: {:ok, User.t}
+  def claim_invite(%{} = params) do
```
We'll need to accommodate invites for projects and general invites to Code Corps.

begedin replied:
Right now, the code sort of does that. It will

- create a user
- join that user with the project, provided the project and role are specified
- assign that user to the assignee

I'm guessing an email will also be involved, which should be pluggable into this process

joshsmith commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,100 @@
+
+  @spec claim_new_user(map) :: {:ok, User.t}
+  defp claim_new_user(%{} = params) do
+    %User{} |> User.registration_changeset(params) |> Repo.insert
```
Not sure if this will change, but in all the 1.6 auto-formatting I've seen everything uses `()`, so `Repo.insert()`.

begedin committed:
Use ProjectUser as provider of valid roles for UserInvite validation

begedin committed:
Remove default "contributor" role from UserInvite

begedin committed:
Add tests for project user invite

joshsmith commented:
> Several questions then
> 
> - will a project admin/owner ever be able to invite an existing codecorps user to their project?
>   - would that not create a `UserInvite` and create something else? Maybe a `ProjectUser` record with a status of `"invited"`?

Yes, they would be able to invite an existing user. But this would be simply a `ProjectUser` with a different `status`, as you suggested. There wouldn't be a need for a `UserInvite` in this case. I'm not sure of a better name to represent the idea, but the invite is to someone who is _not yet_ a user.

> - what happens in the following scenario:
>   - invite is created for a "mike@mail.com"
>   - later, independently, a user registers with that email
>   - even later, that same user tries to claim the invite

Hadn't thought of this. Ideally upon registration we would look through the `UserInvite` table and create any `ProjectUser` records, thereby "claiming" the invite. And again ideally, we would then let a user know that an invite is claimed when they go to a URL with the invite `id` in the query params, and we would also have a `status` for such a claim that differentiates it from a sign up with the invite context.

begedin committed:
Update user requirement on create to prevent invite of existing users

begedin committed:
Add invite claim support to user controller

begedin committed:
Refactor account creation by extracting into Accounts context

begedin committed:
Add policy and tests

begedin committed:
Clean up formatting in UserController

begedin committed:
Fix moduledoc in Policy.ConversationPart

begedin committed:
Add UserInviteView and tests

begedin committed:
Correct Policy.UserInvite behavior

begedin committed:
Add inviter_id support to Plugs.IdsToInteger

begedin committed:
Add UserInvite support to CodeCorps.Policy.can?

begedin committed:
Add UserInviteController with create endpoint and associated tests
- includes tracking of event

begedin committed:
Track UserInvite claim

- required refactoring of Accounts.create and Accounts.UserInvites.claim_invite

begedin commented on lib/code_corps/accounts/accounts.ex:
```diff
@@ -6,20 +6,41 @@ defmodule CodeCorps.Accounts do
+  end
+  def create(%{} = params) do
+    %User{} |> User.registration_changeset(params) |> Repo.insert()
+  end
```
We needed to extract creation into a context. I didn't have a better idea other than this multiple clause approach to differentiate between plain account creation and an invite claim.

begedin commented on lib/code_corps/model/project_user.ex:
```diff
@@ -55,7 +54,11 @@ defmodule CodeCorps.ProjectUser do
+  """
+  @spec roles :: list(String.t())
+  def roles do
+    ~w{pending contributor admin owner}
```
Did an auto-formatting of the file as I was making the `roles` function public (to be used when validating a `UserInvite`

begedin committed:
Format CodeCorps.UserInvite

begedin commented on lib/code_corps_web/controllers/user_controller.ex:
```diff
@@ -28,64 +28,71 @@ defmodule CodeCorpsWeb.UserController do
+    :user_skills
   ]
 
   def preload(data) do
```
Another auto-formatting of file as I was modifying the `create` function. Actual changes are in the function above.

begedin committed:
Move invite claim tracking to controller

begedin committed:
Assume new user when claiming invite

begedin committed:
Add helper to SegmentTraitsBuilder for improved clarity

begedin commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,114 @@
+      [_role, _project_id] ->
+        changeset
+    end
+  end
```
I'm hoping there's a shorter way to write this, but couldn't think of anything.

begedin commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,114 @@
+    |> Changeset.put_assoc(:user, user)
+    |> Changeset.unique_constraint(:project, name: :project_users_user_id_project_id_index)
+    |> Repo.insert()
+  end
```
Technically, validation can fail due to the `unique_constraint`. However, since the only case that can fail is if the user already exists, and the multi will fail on the previous step in that case, this can never happen.

begedin commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,114 @@
+  @spec claim_new_user(map) :: {:ok, User.t()}
+  defp claim_new_user(%{} = params) do
+    %User{} |> User.registration_changeset(params) |> Repo.insert()
+  end
```
As we discussed, we assume new user. 

Existing users will eventually be invitable into the project by creating a special type of `ProjectUser`.

Also, as a user registers for an account directly, not via invite, there should be a step where any existing invites are claimed.

begedin committed:
Switch claimed_invite to has_many

begedin commented on lib/code_corps/accounts/user_invites.ex:
```diff
@@ -0,0 +1,114 @@
+
+  @spec marshall_response(tuple) :: tuple
+  defp marshall_response({:ok, %{user: user, user_invite: user_invite}}) do
+    {:ok, user |> Map.put(:claimed_invites, [user_invite])}
```
This path creates a user by claiming a single invite. This and the normal account creation path should be extended further by adding a step where all other invites are also claimed.

However, I think that's a scope for a different PR.

Right now, we just assume a single invite.

begedin commented on lib/code_corps/model/user.ex:
```diff
@@ -43,6 +43,7 @@ defmodule CodeCorps.User do
     has_one :github_app_installation, CodeCorps.GithubAppInstallation
     has_one :slugged_route, SluggedRoute
 
+    has_many :claimed_invites, CodeCorps.UserInvite, foreign_key: :invitee_id
```
We could add a `:created_invites` to, but it isn't used yet

begedin commented on lib/code_corps/policy/user_invite.ex:
```diff
@@ -0,0 +1,42 @@
+
+    new_role = Map.get(params, "role")
+    do_create?(user_role, new_role)
+  end
```
Two conditions for the case of a project invite

- user is the inviter
- user has a role with authority to invite a user with the specified role

begedin commented on lib/code_corps/policy/user_invite.ex:
```diff
@@ -0,0 +1,42 @@
+
+  def create?(%User{id: user_id}, %{"inviter_id" => inviter_id})
+      when user_id == inviter_id,
+      do: true
```
One condition for a plain invite

- user is the inviter

begedin commented on lib/code_corps/policy/user_invite.ex:
```diff
@@ -0,0 +1,42 @@
+  defp do_create?("owner", "pending"), do: true
+  defp do_create?("owner", "contributor"), do: true
+  defp do_create?("owner", "admin"), do: true
+  defp do_create?(_, _), do: false
```
Could a contributor also invite a "pending" user, or even another contributor?

begedin commented on lib/code_corps_web/controllers/user_controller.ex:
```diff
@@ -28,64 +28,81 @@ defmodule CodeCorpsWeb.UserController do
+
+        _other ->
+          nil
+      end
```
Possibly we can extract this into a `maybe_track_invite_claim`, but other than that, not sure how to improve on this. Note that user creation itself is tracked implicitly via a plug. We should switch that as well at some point, to add clarity.

begedin commented:
@joshsmith I think this is finally in a mergeable state. There are issues to be created to expand this further, pending feedback, but other than that, I think our initial bases are covered. I'll update the PR description with the current state of the PR

begedin committed:
Render 404 on user creation if invite specified but not found

joshsmith commented:
@begedin remember to explicitly re-request a review from me so that it will show up in my reviews.
